katcp/resource_client.py

```diff
@@ -1114,16 +1124,37 @@ def wait(self, sensor_name, condition_or_value, timeout=5):
             sensor is not present
 
         """
+        if quorum is None:
+            quorum = len(self.clients)
+        elif quorum < 1:
+            quorum = int(math.ceil(quorum * len(self.clients)))
+        if timeout and max_grace_period:
+            # Avoid having a grace period longer than or equal to timeout
+            grace_period = min(max_grace_period, timeout / 2.)
+            initial_timeout = timeout - grace_period
+        else:
+            grace_period = max_grace_period
+            initial_timeout = timeout
         # Build dict of futures instead of list as this will be easier to debug
         futures = {}
         for client in self.clients:
-            futures[client.name] = client.wait(sensor_name, condition_or_value,
-                                               timeout)
-        results = yield futures
+            f = client.wait(sensor_name, condition_or_value, initial_timeout)
+            futures[client.name] = f
+        # No timeout required here as all futures will resolve after timeout
+        initial_results = yield until_some(done_at_least=quorum, **futures)
+        results = dict(initial_results)
+        # Identify stragglers and let them all respond within grace period
+        stragglers = {}
+        for client in self.clients:
+            if not results.get(client.name, False):
+                f = client.wait(sensor_name, condition_or_value, grace_period)
+                stragglers[client.name] = f
+        rest_of_results = yield until_some(**stragglers)
+        results.update(dict(rest_of_results))
```

Comment:
Some Tornado newb questions:

Does Tornado just silently drop the futures that haven't resolved yet when their original timeouts expire (we may have exited this method by then)?

We are calling `client.wait` a second time on the same sensor for the stragglers - is KATCP happy to have multiple waits on the same sensor?  I guess the async design would be fine with this.

Assuming a straggler succeeds in the grace period, then both futures will have resolved at that point, right?